File: backend/wallet_keys.py
#!/usr/bin/env python3
"""
Custodial Wallet Key Management for USSD Users
Derives deterministic keypairs from phone numbers for USSD ticket purchases.

Security Model:
- Keys are derived deterministically from phone + master secret
- Master secret stored securely in environment variable
- Same phone always generates same wallet address
- Compatible with web PhoneWallet class (js/zklogin.js)

WARNING: This is a custodial approach. The backend holds user keys.
For web users, prefer zkLogin where users control their own keys.
"""

import os
import hashlib
import hmac
import logging
from typing import Tuple, Optional
from dataclasses import dataclass

logger = logging.getLogger(__name__)

# Master secret for key derivation - MUST be set in production
# This should be a 32+ byte random secret stored securely
MASTER_SECRET = os.environ.get('WALLET_MASTER_SECRET', 'africa-railways-dev-secret-change-in-production')

# Salt for address derivation (must match js/zklogin.js PhoneWallet)
ADDRESS_SALT = 'africa-railways-phone-wallet-v1'

# Salt for private key derivation (different from address salt for security)
PRIVKEY_SALT = 'africa-railways-custodial-privkey-v1'


@dataclass
class CustodialWallet:
    """Represents a custodial wallet derived from phone number"""
    phone_number: str
    normalized_digits: str
    address: str
    private_key_hex: str  # 32 bytes as hex (64 chars)
    

def normalize_phone(phone_number: str) -> str:
    """
    Normalize phone number to 10 digits
    Must match the JavaScript implementation in PhoneWallet.generateWallet()
    """
    # Remove spaces, dashes, parentheses
    cleaned = ''.join(c for c in phone_number if c.isdigit() or c == '+')
    # Get last 10 digits, pad if needed
    digits = cleaned.replace('+', '')[-10:].zfill(10)
    return digits


def derive_address(phone_number: str) -> str:
    """
    Derive SUI wallet address from phone number.
    MUST match the JavaScript PhoneWallet.generateWallet() algorithm exactly.
    
    This ensures the same phone number generates the same address
    whether accessed via USSD, web, or mobile app.
    """
    digits = normalize_phone(phone_number)
    
    # Create deterministic seed (matches JS implementation)
    input_str = digits + ADDRESS_SALT
    
    # SHA-256 hash
    hash_bytes = hashlib.sha256(input_str.encode()).digest()
    
    # Convert to hex address (0x + 64 hex chars)
    address = '0x' + hash_bytes.hex()
    
    return address


def derive_private_key(phone_number: str) -> str:
    """
    Derive private key from phone number using HMAC with master secret.
    
    This is separate from address derivation for security:
    - Address can be computed by anyone with the phone number
    - Private key requires the master secret (backend only)
    """
    digits = normalize_phone(phone_number)
    
    # Use HMAC-SHA256 with master secret for key derivation
    # This ensures only the backend with the secret can derive keys
    key_input = f"{digits}:{PRIVKEY_SALT}"
    
    private_key_bytes = hmac.new(
        MASTER_SECRET.encode(),
        key_input.encode(),
        hashlib.sha256
    ).digest()
    
    return private_key_bytes.hex()


def get_custodial_wallet(phone_number: str) -> CustodialWallet:
    """
    Get or create a custodial wallet for a phone number.
    
    Args:
        phone_number: User's phone number (any format)
    
    Returns:
        CustodialWallet with address and private key
    """
    digits = normalize_phone(phone_number)
    address = derive_address(phone_number)
    private_key = derive_private_key(phone_number)
    
    logger.info(f"🔐 Derived custodial wallet for {phone_number[:6]}***")
    logger.debug(f"   Address: {address[:16]}...")
    
    return CustodialWallet(
        phone_number=phone_number,
        normalized_digits=digits,
        address=address,
        private_key_hex=private_key
    )


def verify_address_matches(phone_number: str, expected_address: str) -> bool:
    """
    Verify that a phone number derives to the expected address.
    Used to validate that a user owns a wallet.
    """
    derived = derive_address(phone_number)
    return derived.lower() == expected_address.lower()


# SUI-specific key conversion
def private_key_to_sui_format(private_key_hex: str) -> str:
    """
    Convert raw private key to SUI keypair format.
    SUI uses Ed25519 keys with a specific encoding.
    
    Returns base64-encoded keypair string for use with pysui.
    """
    import base64
    
    # SUI Ed25519 flag byte
    ED25519_FLAG = 0x00
    
    # Convert hex to bytes
    key_bytes = bytes.fromhex(private_key_hex)
    
    # Prepend flag byte
    sui_key = bytes([ED25519_FLAG]) + key_bytes
    
    # Base64 encode
    return base64.b64encode(sui_key).decode()


# Test function
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    
    print("=" * 60)
    print("🔐 Testing Custodial Wallet Key Derivation")
    print("=" * 60)
    
    # Test with the master wallet phone number
    test_phone = "+260975190740"
    
    print(f"\nPhone: {test_phone}")
    print(f"Normalized: {normalize_phone(test_phone)}")
    
    # Derive wallet
    wallet = get_custodial_wallet(test_phone)
    
    print(f"\nDerived Wallet:")
    print(f"  Address: {wallet.address}")
    print(f"  Private Key: {wallet.private_key_hex[:16]}...{wallet.private_key_hex[-8:]}")
    
    # Verify consistency
    wallet2 = get_custodial_wallet(test_phone)
    assert wallet.address == wallet2.address, "Address derivation not deterministic!"
    assert wallet.private_key_hex == wallet2.private_key_hex, "Key derivation not deterministic!"
    print(f"\n✅ Derivation is deterministic")
    
    # Test different formats of same number
    formats = [
        "+260975190740",
        "260975190740",
        "0975190740",
        "975190740",
        "+260 975 190 740",
        "0975-190-740"
    ]
    
    print(f"\n📱 Testing phone number format normalization:")
    addresses = set()
    for fmt in formats:
        addr = derive_address(fmt)
        addresses.add(addr)
        print(f"  {fmt:20} → {addr[:20]}...")
    
    if len(addresses) == 1:
        print(f"\n✅ All formats derive to same address")
    else:
        print(f"\n❌ WARNING: Different formats gave different addresses!")
    
    print("\n" + "=" * 60)

File: backend/wallet_keys.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"log"
	"os"
	"regexp"
	"strings"
)

// Custodial wallet key management for USSD users
// Derives deterministic keypairs from phone numbers

var (
	// Master secret for key derivation - MUST be set in production
	WalletMasterSecret = getWalletMasterSecret()
	
	// Salt for address derivation (must match js/zklogin.js PhoneWallet)
	AddressSalt = "africa-railways-phone-wallet-v1"
	
	// Salt for private key derivation
	PrivKeySalt = "africa-railways-custodial-privkey-v1"
)

// CustodialWallet represents a wallet derived from phone number
type CustodialWallet struct {
	PhoneNumber      string `json:"phone_number"`
	NormalizedDigits string `json:"normalized_digits"`
	Address          string `json:"address"`
	PrivateKeyHex    string `json:"-"` // Never expose in JSON
}

func getWalletMasterSecret() string {
	secret := os.Getenv("WALLET_MASTER_SECRET")
	if secret == "" {
		log.Println("⚠️ WALLET_MASTER_SECRET not set, using dev default")
		return "africa-railways-dev-secret-change-in-production"
	}
	return secret
}

// NormalizePhone normalizes phone number to 10 digits
// Must match the JavaScript implementation in PhoneWallet.generateWallet()
func NormalizePhone(phoneNumber string) string {
	// Remove all non-digit characters except +
	re := regexp.MustCompile(`[^\d+]`)
	cleaned := re.ReplaceAllString(phoneNumber, "")
	
	// Remove + and get digits
	digits := strings.ReplaceAll(cleaned, "+", "")
	
	// Get last 10 digits, pad if needed
	if len(digits) > 10 {
		digits = digits[len(digits)-10:]
	}
	for len(digits) < 10 {
		digits = "0" + digits
	}
	
	return digits
}

// DeriveAddress derives SUI wallet address from phone number
// MUST match the JavaScript PhoneWallet.generateWallet() algorithm exactly
func DeriveAddress(phoneNumber string) string {
	digits := NormalizePhone(phoneNumber)
	
	// Create deterministic seed (matches JS implementation)
	inputStr := digits + AddressSalt
	
	// SHA-256 hash
	hash := sha256.Sum256([]byte(inputStr))
	
	// Convert to hex address (0x + 64 hex chars)
	return "0x" + hex.EncodeToString(hash[:])
}

// DerivePrivateKey derives private key from phone number using HMAC
func DerivePrivateKey(phoneNumber string) string {
	digits := NormalizePhone(phoneNumber)
	
	// Use HMAC-SHA256 with master secret
	keyInput := fmt.Sprintf("%s:%s", digits, PrivKeySalt)
	
	h := hmac.New(sha256.New, []byte(WalletMasterSecret))
	h.Write([]byte(keyInput))
	privateKeyBytes := h.Sum(nil)
	
	return hex.EncodeToString(privateKeyBytes)
}

// GetCustodialWallet gets or creates a custodial wallet for a phone number
func GetCustodialWallet(phoneNumber string) *CustodialWallet {
	digits := NormalizePhone(phoneNumber)
	address := DeriveAddress(phoneNumber)
	privateKey := DerivePrivateKey(phoneNumber)
	
	log.Printf("🔐 Derived custodial wallet for %s***", phoneNumber[:6])
	
	return &CustodialWallet{
		PhoneNumber:      phoneNumber,
		NormalizedDigits: digits,
		Address:          address,
		PrivateKeyHex:    privateKey,
	}
}

// VerifyAddressMatches verifies that a phone number derives to the expected address
func VerifyAddressMatches(phoneNumber, expectedAddress string) bool {
	derived := DeriveAddress(phoneNumber)
	return strings.EqualFold(derived, expectedAddress)
}

// PrivateKeyToSUIFormat converts raw private key to SUI keypair format
// SUI uses Ed25519 keys with a specific encoding
func PrivateKeyToSUIFormat(privateKeyHex string) (string, error) {
	// SUI Ed25519 flag byte
	const ed25519Flag = 0x00
	
	// Convert hex to bytes
	keyBytes, err := hex.DecodeString(privateKeyHex)
	if err != nil {
		return "", fmt.Errorf("invalid private key hex: %w", err)
	}
	
	// Prepend flag byte
	suiKey := append([]byte{ed25519Flag}, keyBytes...)
	
	// Base64 encode
	return base64.StdEncoding.EncodeToString(suiKey), nil
}

// PaymentSource indicates where the payment request originated
type PaymentSource string

const (
	PaymentSourceUSSD PaymentSource = "ussd"      // Custodial - backend signs
	PaymentSourceWeb  PaymentSource = "web"       // zkLogin - user signs
	PaymentSourceApp  PaymentSource = "app"       // Could be either
)

// DeterminePaymentSource determines how to process payment based on context
func DeterminePaymentSource(walletAddress, phoneNumber string, hasZkProof bool) PaymentSource {
	// If we have a zkLogin proof, user signed via web
	if hasZkProof {
		return PaymentSourceWeb
	}
	
	// If wallet matches phone-derived address, use custodial
	if phoneNumber != "" && VerifyAddressMatches(phoneNumber, walletAddress) {
		return PaymentSourceUSSD
	}
	
	// Default to web (requires user signature)
	return PaymentSourceWeb
}
